src/goto-instrument/contracts/contracts.cpp

```diff
@@ -209,7 +209,10 @@ void code_contractst::check_apply_loop_contracts(
       converter.goto_convert(old_decreases_assignment, havoc_code, mode);
     }
 
-    goto_function.body.destructive_insert(std::next(loop_head), havoc_code);
+    const auto head_loc = loop_head->source_location();
+    while(loop_head->source_location() == head_loc)
+      loop_head = std::next(loop_head);
+    goto_function.body.destructive_insert(loop_head, havoc_code);
```

Comment:
Several issues with both this change and the commit:
1) The commit message fails to answer the "why?" question.
2) There are no tests.
3) Why is the source location a sufficient criterion to select/deselect a particular instruction?

Reply:
I have added a regression test + more description there and also to the commit message.